### PR TITLE
discovery: send service discovery requests in batches

### PR DESCRIPTION
**Pull request by Yumasi**
### What does this PR do?

This PR makes service discovery handle unknown PIDs in batches (default 500) instead of all at once.

### Motivation

DSCVR-567
On machines with very high processes count, service discovery can timeout, causing high kernel CPU usage due to repeating requests.

### Describe how you validated your changes

Covered by unit tests + manual testing with 6000 processes.

### Additional Notes

- Lock/Unlock change around the call to `updateServices` has been removed. Given `updateServices` will go through the batches, it may take a large amount of time (could be in minutes in the case of multi-thousands processes hosts). In order to not block the whole collector, `getCachedProcessData` now returns a snapshot of the `lastCollectedProcesses` field to build the batches, and allow the early release of the lock.

### CODE REVIEW

**Review by gjulianm — APPROVED** [review]

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go [hunk 1]
```diff
@@ -495,22 +589,24 @@ func (c *collector) getProcessEntitiesFromServices(newPids []int32, heartbeatPid
 
 // updateServices retrieves service discovery data for alive processes and returns workloadmeta entities
 func (c *collector) updateServices(alivePids core.PidSet, procs map[int32]*procutil.Process) ([]*workloadmeta.Process, core.PidSet) {
-	newPids, heartbeatPids, pidsToService := c.filterPidsToRequest(alivePids, procs)
+	newPids, heartbeatPids, _ := c.filterPidsToRequest(alivePids, procs)
```
Comment (vitkyrka) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go]: AFAICS this third return value from `filterPidsToRequest` is now never used anymore. Should we remove it?

**Review by vitkyrka — APPROVED** [review]

**Review by s-alad — APPROVED** [review]

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go [hunk 2]
```diff
@@ -387,6 +383,99 @@ func (c *collector) getDiscoveryServices(newPids []int32, heartbeatPids []int32)
...
+	newPids       []int32
+	heartbeatPids []int32
+}
+
+func buildServiceDiscoveryPIDBatches(newPids []int32, heartbeatPids []int32, batchSize int) []serviceDiscoveryPIDBatch {
+	if len(newPids) == 0 && len(heartbeatPids) == 0 {
+		return nil
+	}
+
+	if batchSize == 0 {
```
Comment (kkhor-datadog) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go]: Could make this more defensive by having `batchSize <= 0` in this check

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go [hunk 3]
```diff
@@ -387,6 +383,99 @@ func (c *collector) getDiscoveryServices(newPids []int32, heartbeatPids []int32)
...
+	}
+	for _, pid := range heartbeatPids {
+		addPID(pid, true)
+	}
+
+	if len(current.newPids) > 0 || len(current.heartbeatPids) > 0 {
+		batches = append(batches, current)
+	}
+
+	return batches
```
Comment (kkhor-datadog) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go:440]: Overall this approach seems fine but optionally, you may want to consider using batching logic we already have: https://github.com/DataDog/datadog-agent/blob/7a91254f43d94c2e62497c5388fe652104a0fc48/pkg/process/util/chunking.go#L11

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go [hunk 4]
```diff
@@ -387,6 +383,99 @@ func (c *collector) getDiscoveryServices(newPids []int32, heartbeatPids []int32)
...
+		return
+	}
+
+	dst.Services = append(dst.Services, src.Services...)
+	dst.InjectedPIDs = append(dst.InjectedPIDs, src.InjectedPIDs...)
+	dst.GPUPIDs = append(dst.GPUPIDs, src.GPUPIDs...)
+}
+
+func (c *collector) getDiscoveryServicesBatched(newPids []int32, heartbeatPids []int32) (*model.ServicesResponse, []int32, []int32) {
+	batches := buildServiceDiscoveryPIDBatches(newPids, heartbeatPids, c.systemProbeConfig.GetInt(serviceCollectionBatchSizeConfigKey))
```
Comment (kkhor-datadog) [inline comp/core/workloadmeta/collectors/internal/process/process_collector.go]: This seems like it's getting the config on every collector run. Can this be looked up once at startup, and then stored for reuse?

File: comp/core/workloadmeta/collectors/internal/process/process_service_collector_test.go [hunk 5]
```diff
@@ -109,15 +111,242 @@ func TestFilterPidsToRequest(t *testing.T) {
...
+	batches := buildServiceDiscoveryPIDBatches(newPids, nil, 500)
+
+	require.Len(t, batches, 3)
+	assert.Len(t, batches[0].newPids, 500)
+	assert.Len(t, batches[1].newPids, 500)
+	assert.Len(t, batches[2].newPids, 200)
+	assert.Empty(t, batches[0].heartbeatPids)
+	assert.Empty(t, batches[1].heartbeatPids)
+	assert.Empty(t, batches[2].heartbeatPids)
+}
```
Comment (kkhor-datadog) [inline comp/core/workloadmeta/collectors/internal/process/process_service_collector_test.go:127]: If you want to implement your own batching, it would be helpful to have more comprehensive tests for edge cases e.g. inputs of length 0/1, batch size 0/1

File: comp/core/workloadmeta/collectors/internal/process/process_collector.go (hunk 3 above)
Comment (Yumasi) [reply comp/core/workloadmeta/collectors/internal/process/process_collector.go:440]: I looked at it, and it won't cut it. This chunking assumes a single type of items, and while technically we only send pids, we need to make the distinction between heartbeats & new pids. So, we'd need to add some glue to add this info alongside the pid to store it in PayloadList, and then split it back again to send it to system-probe.

**Review by kkhor-datadog — APPROVED** [review]
